[TTS][Paddle-Lite]add a TTS demo for ARM Linux

SwimmingTiger opened the pull request:
### PR types
Others

### PR changes
Others

### Describe
Add a TTS demo for ARM Linux.

----------------

# PaddleSpeech TTS 文本到语音 ARM Linux Demo

修改自[demos/TTSAndroid](https://github.com/PaddlePaddle/PaddleSpeech/tree/develop/demos/TTSAndroid)，模型也来自该安卓Demo。

使用与安卓Demo版本相同的[Paddle-Lite](https://github.com/PaddlePaddle/Paddle-Lite)推理库（[Paddle-Lite:68b66fd35](https://github.com/SwimmingTiger/Paddle-Lite/releases/tag/68b66fd35)），
该库兼容 Ubuntu 16.04 到 20.04，如果你的发行版与其不兼容，可以自行从源代码编译。

注意 [Paddle-Lite 2.12](https://github.com/PaddlePaddle/Paddle-Lite/releases/tag/v2.12) 与TTS不兼容，无法导出或运行TTS模型，需要使用更新的版本（比如`develop`分支中的代码）。

### 配置

```
git clone -b upstream-pr1 https://github.com/SwimmingTiger/PaddleSpeech.git
cd ./PaddleSpeech/demos/TTSArmLinux
```

打开 [config.sh](https://github.com/SwimmingTiger/PaddleSpeech/blob/upstream-pr1/demos/TTSArmLinux/config.sh) 按需修改配置。

默认编译64位版本，如果要编译32位版本，把`ARM_ABI=armv8`改成`ARM_ABI=armv7hf`。

### 下载Paddle Lite库文件和模型文件

```
./download.sh
```

### 安装依赖

以 Ubuntu 18.04 为例：

```
sudo apt install build-essential cmake libopencv-dev
```

### 编译

```
./build.sh
```

### 运行

```
./run.sh
```

将把[src/main.cpp](https://github.com/SwimmingTiger/PaddleSpeech/blob/upstream-pr1/demos/TTSArmLinux/src/main.cc)里定义在`sentencesToChoose`数组中的十句话转换为`wav`文件，保存在`output`文件夹中。

### 测试环境
我在 RK3588 上用 docker 测的
docker run --name ubuntu-18.04-armhf --privileged -v ~/work:/parent/work --network=host -it armv7/armhf-ubuntu:18.04
docker run --name ubuntu-18.04 --privileged -v ~/work:/parent/work --network=host -it ubuntu:18.04 # host是arm64

macOS里的docker应该也能这样测吧

lzpengyihao committed:
add a TTS demo for ARM Linux

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/PaddleSpeech?pullRequest=2991) <br/>All committers have signed the CLA.

yt605155624 commented:
大神，加微信聊下？PaddleSpeech 用户群搜索 “子龙”

SmileGoat commented on demos/TTSArmLinux/src/Predictor.hpp:
```diff
@@ -0,0 +1,208 @@
+    std::vector<float> wav;
+
+public:
+    bool init(const std::string &AMModelPath, const std::string &VOCModelPath, int cpuThreadNum, const std::string &cpuPowerMode) {
```
统一一下格式？大小写？可以依照一下google c++ coding style

bool Init(....
LoadModel(...

SwimmingTiger replied:
> 统一一下格式？大小写？可以依照一下google c++ coding style
> 
> bool Init(.... LoadModel(...

已经进行了调整

SmileGoat commented on demos/TTSArmLinux/src/Predictor.hpp:
```diff
@@ -0,0 +1,208 @@
+    float inferenceTime = 0;
+    std::shared_ptr<PaddlePredictor> AMPredictor = nullptr;
+    std::shared_ptr<PaddlePredictor> VOCPredictor = nullptr;
+    std::vector<float> wav;
```
移动到下面？

SmileGoat replied:
私有变量命名：float inference_time_;
                          AMP_predictor_;

SmileGoat replied:
AMPredictor 与 VOCPredictor 如果没有共享的需要话，可以是用 unique_ptr;
可以参考一下：
https://www.cyhone.com/articles/right-way-to-use-cpp-smart-pointer/

SwimmingTiger replied:
> AMPredictor 与 VOCPredictor 如果没有共享的需要话，可以是用 unique_ptr; 可以参考一下： https://www.cyhone.com/articles/right-way-to-use-cpp-smart-pointer/

从接口返回的是一个`shared_ptr`，转换成`unique_ptr`不会出问题吗？

```
template <typename ConfigT>
LITE_API std::shared_ptr<PaddlePredictor> CreatePaddlePredictor(const ConfigT&);
```

SwimmingTiger replied:
> 移动到下面？

已完成

SwimmingTiger replied:
> 私有变量命名：float inference_time_; AMP_predictor_;

已经进行了调整

SmileGoat commented on demos/TTSArmLinux/src/main.cc:
```diff
@@ -0,0 +1,71 @@
+    // 009909 但是因为还没到退休年龄，只能掰着指头捱日子。
+    {40, 17, 177, 116, 120, 214, 71, 8, 154, 47, 40, 30, 182, 214, 260, 140, 155, 83, 153, 126, 180, 262, 115, 155, 57, 37, 7, 262, 45, 262, 115, 182, 171, 8, 175, 116, 261, 112},
+    // 009910 这几天雨水不断，人们恨不得待在家里不出门。
+    {262, 44, 151, 74, 182, 82, 240, 177, 213, 37, 184, 40, 202, 180, 175, 52, 154, 55, 71, 54, 37, 186, 40, 42, 40, 7, 261, 10, 151, 77, 153, 74, 37, 186, 39, 183, 154, 52},
```
是否可以传入问题，然后查到这些id，这样使用会更加方便？

SmileGoat commented on demos/TTSArmLinux/src/main.cc:
```diff
@@ -0,0 +1,71 @@
+    {40, 17, 177, 116, 120, 214, 71, 8, 154, 47, 40, 30, 182, 214, 260, 140, 155, 83, 153, 126, 180, 262, 115, 155, 57, 37, 7, 262, 45, 262, 115, 182, 171, 8, 175, 116, 261, 112},
+    // 009910 这几天雨水不断，人们恨不得待在家里不出门。
+    {262, 44, 151, 74, 182, 82, 240, 177, 213, 37, 184, 40, 202, 180, 175, 52, 154, 55, 71, 54, 37, 186, 40, 42, 40, 7, 261, 10, 151, 77, 153, 74, 37, 186, 39, 183, 154, 52},
+};
```
是否可以传入句子，然后查到这些id，这样使用会更加方便？

SwimmingTiger replied:
> 是否可以传入句子，然后查到这些id，这样使用会更加方便？

传入句子会让使用者误以为demo实现了TTS前端，然后传入任意文本，随后失败。所以我设计为传入index。

lzpengyihao committed:
demos/TTSArmLinux: Clarify compatibility issues with Paddle-Lite 2.12

lzpengyihao committed:
demos/TTSArmLinux: adjust the code style to match the project

lzpengyihao committed:
demos/TTSArmLinux: fix WAV size display

lzpengyihao committed:
demos/TTSArmLinux: Eliminate warnings about datatype mismatches

> Paddle-Lite/lite/api/light_api.cc:462 CheckInputValid]  Error input tensor precision type. Input index (0) Tensor name (text) Require precision type (int64_t) Input precision type (float).

lzpengyihao committed:
demos/TTSArmLinux: download.sh add armv7hf library; add MD5 verify

yt605155624 commented:
TODO: 引入 paddlespeech_tts_cpp

yt605155624 commented on demos/TTSArmLinux/download.sh:
```diff
@@ -0,0 +1,54 @@
+    echo '======================='
+}
+
+download 'inference_lite_lib.armlinux.armv8.gcc.with_extra.with_cv.tar.gz' \
```
这两个压缩包我上传到我们的 bos 上了，辛苦换一下链接（记得recheck 一下 md5 值），github 链接下载有一点点慢
1. https://paddlespeech.bj.bcebos.com/demos/TTSArmLinux/inference_lite_lib.armlinux.armv7hf.gcc.with_extra.with_cv.tar.gz
2. https://paddlespeech.bj.bcebos.com/demos/TTSArmLinux/inference_lite_lib.armlinux.armv8.gcc.with_extra.with_cv.tar.gz

yt605155624 commented:
x86 ubuntu 18.04 上报错如下，应该是符合预期的, 因为 gcc/g++ 环境不对
![image](https://user-images.githubusercontent.com/24568452/223047511-c7dc11b8-cadf-491d-a07d-1990ecb2dd5a.png)

lzpengyihao committed:
demos/TTSArmLinux: Update the download links

 The prebuilt libraries has been placed to a BCE BOS server.

lzpengyihao committed:
demos/TTSArmLinux: all scripts use the paths defined in config.sh

lzpengyihao committed:
demos/TTSArmLinux: Remove unnecessary opencv dependencies

Demo does not use OpenCV functions. Even if the Paddle-Lite library enables OpenCV, it does not need to be installed when building the demo.

lzpengyihao committed:
demos/TTSArmLinux: add a clean script

lzpengyihao committed:
demos/TTSArmLinux: update building docs

add steps for building paddle-lite library.

lzpengyihao committed:
demos/TTSArmLinux: fix missing aarch64-linux-gnu-g++ in CentOS 7

In CentOS 7, its aarch64-redhat-linux-g++.

It might be a good idea not to give a compiler name and let cmake find it by itself.

lzpengyihao committed:
demos/TTSArmLinux: minor modification for docs

lzpengyihao committed:
demos/TTSArmLinux: clarify compatibility with Paddle-Lite versions

yt605155624 commented:
在鲲鹏 （CentOS Linux release 7.9.2009 (AltArch)）上按照 README 最下方重新编译 Paddle-Lite 后可以成功运行
![image](https://user-images.githubusercontent.com/24568452/223104207-e825542c-a0f8-49ba-8ef0-21d6f904cfe7.png)

lzpengyihao committed:
demos/TTSArmLinux: save wav in 16-bit PCM format to reduce compatibility issues

lzpengyihao committed:
demos/TTSArmLinux: Fix wav length field value error

yt605155624 reviewed (approved):
LGTM

yt605155624 commented:
![image](https://user-images.githubusercontent.com/24568452/223625576-598e185a-91e7-458c-92e1-8c536b23f104.png)
